src/eg/src/guardian_proof.rs

```diff
@@ -0,0 +1,276 @@
+use std::borrow::Borrow;
+
+use crate::{
+    election_parameters::ElectionParameters,
+    guardian::GuardianIndex,
+    guardian_public_key_info::GuardianPublicKeyInfo,
+    guardian_secret_key::GuardianSecretKey,
+    hash::{eg_h, HValue},
+};
+use anyhow::{ensure, Context, Ok, Result};
+use num_bigint::BigUint;
+use serde::{Deserialize, Serialize};
+use util::{csprng::Csprng, integer_util::to_be_bytes_left_pad};
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PossessionProofChallenge(
+    #[serde(
+        serialize_with = "util::biguint_serde::biguint_serialize",
+        deserialize_with = "util::biguint_serde::biguint_deserialize"
+    )]
+    pub BigUint,
+);
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct PossessionProofResponse(
+    #[serde(
+        serialize_with = "util::biguint_serde::biguint_serialize",
+        deserialize_with = "util::biguint_serde::biguint_deserialize"
+    )]
+    pub BigUint,
+);
+
+/// Proof of Possession for the coefficient in a guardian's secret key.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct GuardianProof {
+    /// Guardian number, 1 <= i <= [`crate::varying_parameters::VaryingParameters::n`].
+    pub i: GuardianIndex,
+    /// Challenge vector
+    pub c: Vec<PossessionProofChallenge>,
+    /// Response vector
+    pub v: Vec<PossessionProofResponse>,
+}
```

Comment:
Some remarks on `GuardianProof`:
- When a verifier validates a public key, they must also check the corresponding proof of possession (cf. Verification 2 in the specification).  It therefore would make sense to explicitly include the proof in   `GuardianPublicKey`. It would be merged with the `CoefficientCommitments`.
- The format of a proof in the specification is a vector of single proofs, i.e. (commitment, challenge, response). In contrast, we have vectors of challenges and responses (the vector of commitments is implicit in the corresponding `GuardianPublicKey`).  
- One could optimise the proof by making a "vector sigma protocol". Here, all first messages are hashed into one challenge that is then used to compute the responses. However, this would require an update of the specification.

Reply:
I have included the coefficient proofs in `GuardianPublicKey` and `GuardianSecretKey`. It is now a vector of single proofs.